torch/cuda/__init__.py

```diff
@@ -498,35 +499,116 @@ def _strtoul(s: str) -> int:
         if not s:
             return -1
         for idx, c in enumerate(s):
-            if not c.isdigit():
+            if not (c.isdigit() or (idx == 0 and c in '+-')):
                 break
             if idx + 1 == len(s):
                 idx += 1
         return int(s[:idx]) if idx > 0 else -1
 
+    def parse_list_with_prefix(prefix):
+        rcs: List[str] = []
+        for elem in var.split(","):
+            # Repeated id results in empty set
+            if elem in rcs:
+                return cast(List[str], [])
+            # Anything other but prefix is ignored
+            if not elem.startswith(prefix):
+                break
+            rcs.append(elem)
+        return rcs
+
+    if var.startswith("GPU-"):
+        return parse_list_with_prefix("GPU-")
+    if var.startswith("MIG-"):
+        return parse_list_with_prefix("MIG-")
     # CUDA_VISIBLE_DEVICES uses something like strtoul
     # which makes `1gpu2,2ampere` is equivalent to `1,2`
     rc: Set[int] = set()
     for elem in var.split(","):
-        rc.add(_strtoul(elem.strip()))
+        x = _strtoul(elem.strip())
+        # Repeated ordinal results in empty set
+        if x in rc:
+            return cast(Set[int], set())
+        # Negative value aborts the sequence
+        if x < 0:
+            break
```

Comment:
Should break at `not 0 <= x < device_count`. You need to query the device count from NVML first. For example, on a machine with 8 GPUs, `CUDA_VISIBLE_DEVICES="0,9,1,2"` should be parsed as `[0]` rather than `[0, 9, 1, 2]`.

Reply:
The idea behind the code is that parsing and validating against NVML results is done separately - if visible devices is invalid to begin with no need to make NVML calls.

Reply:
If we want to validate the device ordinal later in a separate NVML call. The result should be ordered. Otherwise, the example case I shown above will produce wrong results.